File: Java/quicksort/llama_3_1_8b.java
import java.util.ArrayList;
import java.util.List;

public class llama_3_1_8b {

    /**
     * This method implements the quick sort algorithm.
     * It recursively sorts the array by selecting a pivot element, partitioning the array around the pivot, and then recursively sorting the sub-arrays.
     * 
     * @param arr The array to be sorted.
     * @param low The starting index of the array.
     * @param high The ending index of the array.
     */
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(arr, low, high);
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
    }

    /**
     * This method partitions the array around a pivot element.
     * It rearranges the array such that all elements less than the pivot are on the left of the pivot, and all elements greater than the pivot are on the right.
     * 
     * @param arr The array to be partitioned.
     * @param low The starting index of the array.
     * @param high The ending index of the array.
     * @return The index of the pivot element after partitioning.
     */
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    /**
     * This method swaps two elements in the array.
     * 
     * @param arr The array containing the elements to be swapped.
     * @param i The index of the first element.
     * @param j The index of the second element.
     */
    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

 public static void main(String[] args) {
        List<Integer> input = new ArrayList<>();

        try (java.io.BufferedReader br = new java.io.BufferedReader(new java.io.FileReader("../../inputs.txt"))) {
            String line;
            while ((line = br.readLine()) != null) {
                input.add(Integer.parseInt(line.trim()));
            }
        } catch (java.io.IOException e) {
            return;
        }

        // Converte List<Integer> diretamente para int[]
        int[] arr = input.stream().mapToInt(Integer::intValue).toArray();

        // Usa o método existente
        quickSort(arr, 0, arr.length - 1);

    }

}

File: Python/quicksort/llama-3-1-8b.py

def quick_sort(arr):
    """
    Optimized implementation of the quick sort algorithm.

    This function uses the 'Lomuto' partition scheme, which is a variation of the standard 'Hoare' partition scheme.
    It is slightly faster and more efficient in practice.

    :param arr: The input array to be sorted.
    :return: The sorted array.
    """
    _quick_sort_helper(arr, 0, len(arr) - 1)

def _quick_sort_helper(arr, low, high):
    """
    Recursive helper function for the quick sort algorithm.

    :param arr: The input array to be sorted.
    :param low: The low index of the current partition.
    :param high: The high index of the current partition.
    """
    if low < high:
        # Partition the array and get the pivot index
        pivot_index = _partition(arr, low, high)

        # Recursively sort the sub-arrays on either side of the pivot
        _quick_sort_helper(arr, low, pivot_index - 1)
        _quick_sort_helper(arr, pivot_index + 1, high)


def _partition(arr, low, high):
    """
    Partition the array around a pivot element.

    This function uses the 'Lomuto' partition scheme.

    :param arr: The input array to be partitioned.
    :param low: The low index of the current partition.
    :param high: The high index of the current partition.
    :return: The index of the pivot element.
    """
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


# Example usage:
import sys

def main():
    input_path = "inputs.txt" if len(sys.argv) < 2 else sys.argv[1]

    try:
        with open(input_path, 'r') as f:
            arr = [int(line.strip()) for line in f if line.strip()]
    except FileNotFoundError:
        return

    sorted_arr = quick_sort(arr)
    

if __name__ == "__main__":
    main()
